VirtoCommerce.Platform.Data/Assets/AssetEntryService.cs

```diff
@@ -0,0 +1,152 @@
+using VirtoCommerce.Platform.Core.Assets;
+using VirtoCommerce.Platform.Core.Common;
+using VirtoCommerce.Platform.Data.Caching;
+using VirtoCommerce.Platform.Data.Repositories;
+
+namespace VirtoCommerce.Platform.Data.Assets
+{
+    public class AssetEntryService : IAssetEntryService, IAssetEntrySearchService
+    {
+        private readonly Func<IPlatformRepository> _platformRepository;
+        private readonly IBlobUrlResolver _blobUrlResolver;
+
+        public AssetEntryService(Func<IPlatformRepository> repositoryFactory, IBlobUrlResolver blobUrlResolver)
+        {
+            _platformRepository = repositoryFactory;
+            _blobUrlResolver = blobUrlResolver;
+        }
+
+        public async Task<GenericSearchResult<AssetEntry>> SearchAssetEntriesAsync(AssetEntrySearchCriteria criteria)
+        {
+            criteria = criteria ?? AbstractTypeFactory<AssetEntrySearchCriteria>.TryCreateInstance();
+
+            using (var repository = _platformRepository())
+            {
+                var query =  repository.AssetEntries;
+
+                if (!string.IsNullOrEmpty(criteria.SearchPhrase))
+                {
+                    query = query.Where(x =>
+                        x.Name.Contains(criteria.SearchPhrase) || x.RelativeUrl.Contains(criteria.SearchPhrase));
+                }
+
+                if (!string.IsNullOrEmpty(criteria.LanguageCode))
+                {
+                    query = query.Where(x => x.LanguageCode == criteria.LanguageCode);
+                }
+
+                if (!string.IsNullOrEmpty(criteria.Group))
+                {
+                    query = query.Where(x => x.Group == criteria.Group);
+                }
+
+                if (!criteria.Tenants.IsNullOrEmpty())
+                {
+                    var tenants = criteria.Tenants.Where(x => x.IsValid).ToArray();
+                    if (tenants.Any())
+                    {
+                        var tenantsStrings = tenants.Select(x => x.ToString());
+                        query = query.Where(x => tenantsStrings.Contains(x.TenantId + "_" + x.TenantType));
+                    }
+                }
+
+                var result = new GenericSearchResult<AssetEntry>()
+                {
+                    TotalCount = query.Count()
+                };
+
+                var sortInfos = criteria.SortInfos;
+                if (sortInfos.IsNullOrEmpty())
+                {
+                    sortInfos = new[]
+                    {
+                        new SortInfo
+                        {
+                            SortColumn = "CreatedDate",
+                            SortDirection = SortDirection.Descending
+                        }
+                    };
+                }
+
+                query = query.OrderBySortInfos(sortInfos);
+
+                var ids = query
+                    .Skip(criteria.Skip)
+                    .Take(criteria.Take)
+                    .Select(x => x.Id).ToList();
```

Comment:
ToListAsync()

Reply:
linq does not have asynchronous requests